Add streaming support

shinfan opened the pull request:
cc/ @jmuk 

1. The following new streaming classes are added:
  - StreamingApiCallable
    - Public access
    - Callable object which supports client/server/bidi streaming calls with the given channel
  - StreamingCallable
    - Package private (used by StreamingApiCallable)
    - Bridge between gax to actual grpc streaming calls.
  - StreamingCallSettings
    - Public access
    - Creates the StreamingAPICallable object for the method based on the MethodDescriptor object from the generated gRPC code 

2. The original non-streaming API callable/settings classes are renamed with “Unary” prefix to be distinguished from the new streaming classes:

  - ApiCallable -> UnaryApiCallable
  - ApiCallSettings -> UnaryApiCallSettings
  - ApiCallSettingsTyped -> UnaryApiCallSettingsTyped

3. Refactored CallContext class to only contain CallOptions and Channel. The benefits of this refactoring are:

  - Remove the type parameter from CallContext to make the code cleaner.
  - Share CallContext for both streaming and non-streaming API calls. 

4. Removed ChannelBindingCallable. Handles channel binding within UnaryApiCallable instead.

pongad commented on src/main/java/com/google/api/gax/grpc/StreamingApiCallable.java:
```diff
@@ -0,0 +1,99 @@
+  private final StreamingCallable<RequestT, ResponseT> callable;
+  private ManagedChannel channel;
+
+  public StreamingApiCallable(StreamingCallable<RequestT, ResponseT> callable) {
```
Should this be public? StreamingCallable is package-private. If no one else can use it anyway, I think we might as well just make this package-private too.

shinfan replied:
You are right. Changed it to package-private.

shinfan commented:
Updated unit tests. PTAL.

pongad commented:
@shinfan I made a comment on the design doc. Would be good if we can address it, but we can do it later as well.

garrettjonesgoogle commented:
As discussed offline, after seeing the changes in context, I think we should go for UnaryApi* instead of FutureApi*.

shinfan commented:
PTAL

@garrettjonesgoogle I have done the renaming.

@pongad I have addressed the comment by adding an iteration streaming method to StreamingApiCallable

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/ApiException.java:
```diff
@@ -39,7 +39,7 @@
  *
  * <p>
- * It stores information useful for functionalities in {@link ApiCallable}. For more information
+ * It stores information useful for functionalities in {@link UnaryApiCallable}. For more information
```
We use it for StreamingApiCallable too, right?

shinfan replied:
Done

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/BundlingCallSettings.java:
```diff
@@ -43,21 +43,21 @@
 
 /**
- * A settings class to configure an ApiCallable for calls to an API method that supports
+ * A settings class to configure an UnaryApiCallable for calls to an API method that supports
```
s/an/a/

Do a search through all of the code to replace all of the cases. Note, there is at least one capitalized case (s/An/A).

shinfan replied:
Done

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/CallContext.java:
```diff
@@ -43,22 +44,36 @@
-  public static <T> CallContext<T> of(Channel channel, CallOptions callOptions, T request) {
-    return new CallContext<T>(channel, callOptions, request);
+  public static <RequestT, ResponseT> CallContext<RequestT, ResponseT> of(
+      Channel channel, StreamObserver<ResponseT> responseObserver, RequestT request) {
```
We're getting into a potential combinatorial explosion here. Might someone want to provide both CallOptions and StreamObserver?

Also, what should we do if someone provides us with a StreamObserver in a future call?

Is it worth using CallContext for both types of calls? (Maybe it is, but we should make sure we know why)

shinfan replied:
Refactored the request/observer out of the context as discussed offiline.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/PageStreamingCallable.java:
```diff
@@ -60,9 +60,11 @@ public String toString() {
+        CallContext.of(context.getChannel(), context.getCallOptions(), context.getRequest());
     PagedListResponse<RequestT, ResponseT, ResourceT> pagedListResponse =
-        new PagedListResponseImpl<>(callable, pageDescriptor, context);
+        new PagedListResponseImpl<>(callable, pageDescriptor, newContext);
```
Why are you creating a new context? It's not a proper copy since it is skipping StreamingObserver.

shinfan replied:
The CallContext object is no longer parameterized so this is reverted.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/StreamingApiCallable.java:
```diff
@@ -0,0 +1,135 @@
+ * An UnaryApiCallable is an immutable object which is capable of making RPC calls to streaming
+ * API methods.
+ *
+ * <p>It is considered advanced usage for a user to create an StreamingApiCallable themselves. This
```
s/an/a/

shinfan replied:
Done. Fixed the same typo in UnaryApiCallable too.

garrettjonesgoogle commented on src/test/java/com/google/api/gax/grpc/StreamingApiCallableTest.java:
```diff
@@ -0,0 +1,171 @@
+@RunWith(JUnit4.class)
+public class StreamingApiCallableTest {
+
+  private static class StashCallable<RequestT, ResponseT>
```
Maybe put this private class at the end instead of the beginning.

shinfan replied:
Done.

shinfan commented:
@garrettjonesgoogle I have refactored the CallContext and addressed the comments. PTAL.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/CallContext.java:
```diff
@@ -35,31 +35,26 @@
-  public static <T> CallContext<T> of(Channel channel, CallOptions callOptions, T request) {
-    return new CallContext<T>(channel, callOptions, request);
-  }
+  public static final CallContext DEFAULT = new CallContext(null, CallOptions.DEFAULT);
```
The Java peoples recommended that we make a static accessor function even in cases like this.

shinfan replied:
Done.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/StreamingApiCallable.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.Iterator;
+
+/**
+ * A UnaryApiCallable is an immutable object which is capable of making RPC calls to streaming
```
Wrong class name here

shinfan replied:
Fixed

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/StreamingApiCallable.java:
```diff
@@ -0,0 +1,132 @@
+   * @param responseObserver {@link io.grpc.stub.StreamObserver} to observe the streaming responses
+   * @param request request
+   */
+  public void serverStreamingCall(StreamObserver<ResponseT> responseObserver, RequestT request) {
```
It seems more natural (to me) for the request to be first (like gRPC has it)

shinfan replied:
Done

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/StreamingCallable.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ * <p>Package-private for internal use.
+ */
+class StreamingCallable<RequestT, ResponseT> {
```
Analogous to the unary case, call this one DirectStreamingCallable.

shinfan replied:
Done

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/StreamingApiCallable.java:
```diff
@@ -0,0 +1,132 @@
+ * instances of StreamingCallSettings.Builder which are exposed through the API wrapper class's
+ * settings class.
+ */
+public class StreamingApiCallable<RequestT, ResponseT> {
```
You don't provide any means for the caller to provide a CallContext.

shinfan replied:
I added a context variant for each of the streaming call method. Done.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/StreamingApiCallable.java:
```diff
@@ -87,19 +85,57 @@ public void bind(ManagedChannel channel) {
+      StreamObserver<ResponseT> responseObserver, CallContext context) {
+    if (context.getChannel() == null) {
+      context = context.withChannel(channel);
+    }
```
For ApiCallable, the ChannelBindingCallable class does this work. I think we should follow the same model for Streaming too. I think this means you will need a StreamingCallable interface akin to FutureCallable, which will have subclasses ChannelBindingStreamingCallable and DirectStreamingCallable.

shinfan replied:
I can understand why the unary callable has this structure since they have many functionalities such as retrying and bundling. But for streaming I don't see much benefits of having an extra channel binding class while we can handle channel binding directly in one class.

garrettjonesgoogle replied:
Well actually, the unary case could to it exactly as you do, and bind the channel in UnaryApiCallable, because it is the top-level entry point, and there is no reason for the channel to change once the call has been issued. I would rather you make them consistent in this regard than do one differently.

shinfan replied:
Cool, I removed the ChannelBindingCallable for consistency.

garrettjonesgoogle commented:
I will approve this once the GAX version is bumped separate from this PR.

shinfan commented:
Version bump PR merged: https://github.com/googleapis/gax-java/pull/114

garrettjonesgoogle commented:
Ok, this PR LGTM
